### PR TITLE
Add interpolate operator replace bilinear_interp_op and add nearest neighbor interp mode

### PR DESCRIPTION
**Pull request by heavengate**

### CODE REVIEW

File: paddle/fluid/operators/nearest_neighbor_interp_op.cc [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+};
+
+class NearestNeighborInterpOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X",
+             "The input tensor of nearest neighbor interpolation, "
+             "This is a 4-D tensor with shape of (N x C x h x w)");
+    AddInput("OutSize",
+             "This is a 1-D tensor with two number. "
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/nearest_neighbor_interp_op.cc]: number-> numbers

File: paddle/fluid/operators/nearest_neighbor_interp_op.cc [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+             "The first number is height and the second number is width.")
+        .AsDispensable();
+    AddOutput("Out", "The dimension of output is (N x C x out_h x out_w)");
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/nearest_neighbor_interp_op.cc]: The shape of output is .... ?

File: paddle/fluid/operators/nearest_neighbor_interp_op.cc [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+REGISTER_OPERATOR(nearest_neighbor_interp, ops::NearestNeighborInterpOp,
+                  ops::NearestNeighborInterpOpMaker,
+                  paddle::framework::DefaultGradOpDescMaker<true>);
+REGISTER_OPERATOR(nearest_neighbor_interp_grad,
+                  ops::NearestNeighborInterpOpGrad);
+REGISTER_OP_CPU_KERNEL(nearest_neighbor_interp,
+                       ops::NearestNeighborInterpKernel<float>,
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/nearest_neighbor_interp_op.cc]: 没注册下double类型是因为`atomicAdd`不支持double?

File: paddle/fluid/operators/nearest_neighbor_interp_op.cu [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  }
+}
+
+template <typename T>
+__global__ void KeNearestNeighborInterpBw(
+    T* in, const size_t in_img_h, const size_t in_img_w, const size_t input_h,
+    const size_t input_w, const T* out, const size_t out_img_h,
+    const size_t out_img_w, const size_t output_h, const size_t output_w,
+    const size_t num_channels, const T ratio_h, const T ratio_w) {
+  int nthreads = output_h * output_w;
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/nearest_neighbor_interp_op.cu]: 如果我们遍历input_h*input_w, 分别收集input[i][j]的gradients，是不是就可以不用atomicAdd了？

File: paddle/fluid/operators/nearest_neighbor_interp_op.cu [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  }
+}
+
+template <typename T>
+class NearestNeighborInterpOpCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    PADDLE_ENFORCE(platform::is_gpu_place(ctx.GetPlace()),
+                   "This kernel only runs on GPU device.");
+    auto* input = ctx.Input<Tensor>("X");      // float tensor
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/nearest_neighbor_interp_op.cu]: remove `// float tensor` ?

File: paddle/fluid/operators/nearest_neighbor_interp_op.cu [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    int in_hw = in_h * in_w;
+    int out_hw = out_h * out_w;
+    int in_chw = c * in_hw;
+    int out_chw = c * out_hw;
+
+    T ratio_h = (out_h > 1) ? static_cast<T>(in_h - 1) / (out_h - 1) : 0.f;
+    T ratio_w = (out_w > 1) ? static_cast<T>(in_w - 1) / (out_w - 1) : 0.f;
+
+    if (in_h == out_h && in_w == out_w) {
+      memcpy(output_data, input_data, input->numel() * sizeof(T));
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/nearest_neighbor_interp_op.cu]: 建议这里直接用TensorCopy。

**Review by wanghaoshuang — COMMENTED** [review]
resize_nearest op和resize_bilinear op有大量代码重复，可以考虑合并一下。
如果比较着急用，可以先merge.

File: paddle/fluid/operators/nearest_neighbor_interp_op.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/operators/nearest_neighbor_interp_op.cc]: Done

File: paddle/fluid/operators/nearest_neighbor_interp_op.cc (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/operators/nearest_neighbor_interp_op.cc]: Done

File: paddle/fluid/operators/nearest_neighbor_interp_op.cc (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/operators/nearest_neighbor_interp_op.cc]: I have changed to use platform::CudaAtomicAdd which can support double/float/int and add double and int support in op register

File: paddle/fluid/operators/nearest_neighbor_interp_op.cu (hunk 4 above)
Comment (heavengate) [reply paddle/fluid/operators/nearest_neighbor_interp_op.cu]: It is said to be better that adjacent threads access adjacent data in CUDA program optimization, in this case, maybe atomic add is inevitable. Currently, threads parallel in each pixel, I'll try to redcue parallel thread numbers in following commit.

File: paddle/fluid/operators/nearest_neighbor_interp_op.cu (hunk 5 above)
Comment (heavengate) [reply paddle/fluid/operators/nearest_neighbor_interp_op.cu]: Done

File: paddle/fluid/operators/nearest_neighbor_interp_op.cu (hunk 6 above)
Comment (heavengate) [reply paddle/fluid/operators/nearest_neighbor_interp_op.cu]: Done

**Review by heavengate — COMMENTED** [review]
Thanks for your reviewing and have done for all suggestions.

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -5672,6 +5678,29 @@ def resize_bilinear(input, out_shape=None, scale=None, name=None):
     return image_resize(input, out_shape, scale, name, 'BILINEAR')
 
 
+@templatedoc(op_type="interpolate")
+def resize_nearest(input, out_shape=None, scale=None, name=None):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: 会不会和resize_bilinear存在同样的问题，需要考虑动态shape的问题。

File: python/paddle/fluid/layers/nn.py [hunk 8]
```diff
@@ -5672,6 +5678,29 @@ def resize_bilinear(input, out_shape=None, scale=None, name=None):
...
+    """
+    ${comment}
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: 因为和resize_bilinear公用的op， 这样写，文档会和resize_bilinear一样吧。

更倾向于这里comments写出来， 明确告诉用是nearest插值，下同。

File: python/paddle/fluid/layers/nn.py (hunk 7 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done.

File: python/paddle/fluid/layers/nn.py (hunk 8 above)
Comment (heavengate) [reply python/paddle/fluid/layers/nn.py]: Done.

**Review by heavengate — COMMENTED** [review]
Thanks for reviewing and have done for all suggestions.

File: paddle/fluid/API.spec [hunk 9]
```diff
@@ -118,9 +118,10 @@ paddle.fluid.layers.label_smooth ArgSpec(args=['label', 'prior_dist', 'epsilon',
 paddle.fluid.layers.roi_pool ArgSpec(args=['input', 'rois', 'pooled_height', 'pooled_width', 'spatial_scale'], varargs=None, keywords=None, defaults=(1, 1, 1.0))
 paddle.fluid.layers.roi_align ArgSpec(args=['input', 'rois', 'pooled_height', 'pooled_width', 'spatial_scale', 'sampling_ratio', 'name'], varargs=None, keywords=None, defaults=(1, 1, 1.0, -1, None))
 paddle.fluid.layers.dice_loss ArgSpec(args=['input', 'label', 'epsilon'], varargs=None, keywords=None, defaults=(1e-05,))
-paddle.fluid.layers.image_resize ArgSpec(args=['input', 'out_shape', 'scale', 'name', 'resample'], varargs=None, keywords=None, defaults=(None, None, None, 'BILINEAR'))
+paddle.fluid.layers.image_resize ArgSpec(args=['input', 'out_shape', 'scale', 'name', 'resample', 'actual_shape'], varargs=None, keywords=None, defaults=(None, None, None, 'BILINEAR', None))
```
Comment (panyx0718) [inline paddle/fluid/API.spec:121]: what's the difference between this `image_resize` and the following `image_resize_xxx` operators. Is image_resize using one algorithm of the following APIs?

Comment (heavengate) [reply paddle/fluid/API.spec:121]: image_resize is the basic API with parameters such as out_shape, scale, resample, following APIs use image_resize, image_resize_short resize the short edge, resize_nearest uses nearest neighbor interpolation resample method and resize_bilinear uses bilinear interpolation.

File: paddle/fluid/API.spec [hunk 10]
```diff
@@ -118,9 +118,10 @@ paddle.fluid.layers.label_smooth ArgSpec(args=['label', 'prior_dist', 'epsilon',
...
 paddle.fluid.layers.image_resize_short ArgSpec(args=['input', 'out_short_len', 'resample'], varargs=None, keywords=None, defaults=('BILINEAR',))
-paddle.fluid.layers.resize_bilinear ArgSpec(args=['input', 'out_shape', 'scale', 'name'], varargs=None, keywords=None, defaults=(None, None, None))
+paddle.fluid.layers.resize_bilinear ArgSpec(args=['input', 'out_shape', 'scale', 'name', 'actual_shape'], varargs=None, keywords=None, defaults=(None, None, None, None))
+paddle.fluid.layers.resize_nearest ArgSpec(args=['input', 'out_shape', 'scale', 'name', 'actual_shape'], varargs=None, keywords=None, defaults=(None, None, None, None))
```
Comment (panyx0718) [inline paddle/fluid/API.spec:124]: image_resize_xxx is not ideal. But for consistent and easier look up, it seems slightly better to call them image_resize_xxx as well.

**Review by panyx0718 — APPROVED** [review]
LG API

**Review by qingqing01 — APPROVED** [review]

**Comment by zzhzz** [comment]
I try to pull a new operator and my pull request #14404 has been blocked by interpolate operator unittest when PR_CI build. Please help me fix the problem. The error log is here http://ci.paddlepaddle.org/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=28317&_focus=22705

**Comment by zzhzz** [comment]
I try to pull a new operator and my pull request #14404 has been blocked by interpolate operator unittest when PR_CI system checking. Please help me fix this problem. The PR_CI error log is here:
http://ci.paddlepaddle.org/viewLog.html?buildId=28317&buildTypeId=Paddle_PrCi&tab=buildLog&_focus=22087 . thanks

**Comment by heavengate** [comment]
Thanks for your contribution!
  I have rerun the failed CI and this error seems not to recur, maybe GPU was unstable or crashed in your last CI build, the new error is "437 - test_label_semantic_roles (Failed)”, responsible developer has been mentioned to have a look, you could try to rerun CI or we will rerun it after fix the new error.
  Thank you!

  http://ci.paddlepaddle.org/viewLog.html?buildId=28332&buildTypeId=Paddle_PrCi

发件人: zhaozhehao <notifications@github.com<mailto:notifications@github.com>>
答复: PaddlePaddle/Paddle <reply@reply.github.com<mailto:reply@reply.github.com>>
日期: 2018年11月14日 星期三 上午11:49
至: PaddlePaddle/Paddle <Paddle@noreply.github.com<mailto:Paddle@noreply.github.com>>
抄送: Kaipeng Deng <dengkaipeng@baidu.com<mailto:dengkaipeng@baidu.com>>, State change <state_change@noreply.github.com<mailto:state_change@noreply.github.com>>
主题: Re: [PaddlePaddle/Paddle] Add interpolate operator replace bilinear_interp_op and add nearest neighbor interp mode (#14205)


I try to pull a new operator and my pull request #14404<https://github.com/PaddlePaddle/Paddle/pull/14404> has been blocked by interpolate operator unittest when PR_CI system checking. Please help me fix this problem. The PR_CI error log is here:
http://ci.paddlepaddle.org/viewLog.html?buildId=28317&buildTypeId=Paddle_PrCi&tab=buildLog&_focus=22087 . thanks

―

**Pull request merged**